server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1175,80 +1141,39 @@ private static Collection<String> innerResolveAll(Context context) {
          * ultimately returned, instead of the alias or datastream name</li>
          * </ol>
          */
-        private static Collection<String> innerResolve(Context context, List<String> expressions) {
-            if (Objects.requireNonNull(expressions).isEmpty()) {
-                throw new IllegalStateException("Cannot resolve empty index expression");
+        public static Collection<String> resolve(Context context, List<String> expressions) {
+            ExpressionIterable expressionIterable = new ExpressionIterable(context, expressions);
+            // fast exist if there are no wildcards to evaluate
+            if (expressionIterable.hasWildcard() == false) {
+                return expressions;
```

Comment:
There's now a preprocessing step that goes over the elements of the expression in search of the first wildcard. If none is found, no wildcard needs to be resolved, and the resolver moves on (this is equivalent to the original optimisation but more succintly expressed). If a wildcard is found, it's position in the expression list is later reused.